### PR TITLE
Refactor environment event emitter

### PR DESCRIPTION
**Pull request by pattyshack**
1. move event.go & event_test.go from handler into environment
2. renamed (h *EventHandler) to (emitter *eventEmitter)
3. renamed (e *EventCollection) to (collection *EventCollection)
4. removed Child & Merge from EventCollection (dead code)
5. added EventEmitter interface and NoEventEmitter for scripts
6. merged TransactionEnv.EmitEvent's tracing / metering logic into
   eventEmitter.EmitEvent.

### CODE REVIEW

File: fvm/environment/event_emitter.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	}
+
+	payload, err := jsoncdc.Encode(event)
+	if err != nil {
+		return errors.NewEncodingFailuref(
+			"failed to json encode a cadence event: %w",
+			err)
+	}
+
+	payloadSize := uint64(len(payload))
```
Comment (Tonix517) [inline fvm/environment/event_emitter.go:102]: q: do we consider using `json.Compact()` here? is `eventCollectionByteSizeLimit` for compacted or non-compacted json data, or is it really a don't care?

**Review by Tonix517 — APPROVED** [review]
LGTM. just a question out of curiosity

File: fvm/environment/event_emitter.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+// with the runtime environment interface.
+type EventEmitter interface {
+	// Cadence's runtime API.  Note that the script variant will return
+	// OperationNotSupportedError.
+	EmitEvent(event cadence.Event) error
+
+	Events() []flow.Event
+	ServiceEvents() []flow.Event
+}
+
```
Comment (zhangchiqing) [inline fvm/environment/event_emitter.go:29]:
```suggestion
// Ensure the methods of the interface are implemented.
var _ EventEmitter = (*NoEventEmitter)(nil)
var _ EventEmitter = (*EventCollection)(nil)
```

File: fvm/environment/event_emitter.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+type NoEventEmitter struct{}
+
+func (NoEventEmitter) EmitEvent(event cadence.Event) error {
```
Comment (zhangchiqing) [inline fvm/environment/event_emitter.go:32]:
```suggestion
// NoEventEmitter is usually used in the environment for script execution, where no event should be emitted.
func (NoEventEmitter) EmitEvent(event cadence.Event) error {
```

**Review by zhangchiqing — APPROVED** [review]
Looks good. Just had some minor suggestion

**Review by janezpodhostnik — APPROVED** [review]

File: fvm/environment/event_emitter.go (hunk 1 above)
Comment (pattyshack) [reply fvm/environment/event_emitter.go:102]: my guess is they just needed some form of limiting.  the exact value doesn't matter

**Comment by pattyshack** [comment]
bors merge

File: fvm/environment/event_emitter.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+var _ EventEmitter = NoEventEmitter{}
+
+// NoEventEmitter is usually used in the environment for script execution,
+// where no event should be emitted.
+type NoEventEmitter struct{}
+
+func (NoEventEmitter) EmitEvent(event cadence.Event) error {
+	return errors.NewOperationNotSupportedError("EmitEvent")
```
Comment (bluesign) [inline fvm/environment/event_emitter.go:37]: Any chance to make this non-error? This error is limiting a lot of good stuff. ref: https://github.com/onflow/flow-go/issues/2813

**Comment by pattyshack** [comment]
The goal of this PR is to preserve the existing behavior.  I'll let you
guys figure out what the desired behavior should be.

**Pull request merged**